pss/outbox/outbox.go

```diff
@@ -0,0 +1,130 @@
+package outbox
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/pss/message"
+)
+
+type Config struct {
+	NumberSlots int
+	NumWorkers  int
+	Forward     forwardFunction
+}
+
+type Outbox struct {
+	forwardFunc forwardFunction
+	queue       []*outboxMsg
+	slots       chan int
+	process     chan int
+	workers     chan int
+	stopC       chan struct{}
+}
+
+type forwardFunction func(msg *message.Message) error
+
+var ErrOutboxFull = errors.New("outbox full")
+
+const defaultOutboxWorkers = 100
+
+func NewOutbox(config *Config) *Outbox {
+	if config.NumWorkers == 0 {
+		config.NumWorkers = defaultOutboxWorkers
+	}
+	outbox := &Outbox{
+		forwardFunc: config.Forward,
+		queue:       make([]*outboxMsg, config.NumberSlots),
+		slots:       make(chan int, config.NumberSlots),
+		process:     make(chan int),
+		workers:     make(chan int, config.NumWorkers),
+		stopC:       make(chan struct{}),
+	}
+	// fill up outbox slots
+	for i := 0; i < cap(outbox.slots); i++ {
+		outbox.slots <- i
+	}
+	return outbox
+}
+
+func (o *Outbox) Start() {
+	log.Info("Starting outbox")
+	go o.processOutbox()
+}
+
+func (o *Outbox) Stop() {
+	log.Info("Stopping outbox")
+	close(o.stopC)
+}
+
+// Enqueue a new element in the outbox if there is any slot available.
+// Then send it to process. This method is blocking in the process channel!
```

Comment:
`process channel`: internal implementation detail in comment to the exposed method.